File: dft.py
import numpy as np

def DFT(x):
    N = len(x)
    X = []

    for k in range(N):
        sum_val = 0 + 0j  # complex number

        for n in range(N):
            angle = -2 * np.pi * k * n / N
            w = complex(np.cos(angle), np.sin(angle))

            sum_val += x[n] * w

        X.append(sum_val)

    return X


# main
N = int(input("no. of samples: "))

print("enter signal samples:")
x = list(map(float, input().split()))

X = DFT(x)

print("performed DFT of the given signal:")
for k in range(N):
    print(f"X[{k}] = {X[k]}")

for k in range (N):
    print(f"magnitude of X[{k}] = {abs(X[k])}")

File: DFT.cpp
#define _USE_MATH_DEFINES

#include <vector>
#include <bits/stdc++.h>
#include <cmath>

using namespace std;

/*
    DFT FORMULA:
    X[m] = sum_{n=0}^{N-1}  x[n] * e^(-j*2*pi*m*n/N)

    where:
    - X[m] is the m-th frequency component
    - x[n] is the n-th time-domain sample
    - N is the total number of samples
    - represent in cos sin also
*/

vector<complex<double>> DFT(vector<double> x){
    int n=x.size();
    vector<complex<double>> X(n);
    for(int i=0;i<n;i++){
        complex<double> sum(0,0);
        for(int j=0;j<n;j++){
            double angle=-2*M_PI*i*j/n;
            complex<double> w(cos(angle),sin(angle));
            sum+=x[j]*w;
        }
        X[i]=sum;
    }
    return X;
}

int main(){
    int n;
    cout<<"no. of samples:"<<endl;
    cin>>n;
    vector<double> sig(n);
    cout<<"enter signal samples:"<<endl;
    for(int i=0;i<n;i++){
        cin>>sig[i];
    }
    vector<complex<double>> X = DFT(sig);
    cout<<"performed DFT of the given signal:"<<endl;
    for(int i=0;i<n;i++){
        cout<<"X["<<i<<"] = "<<X[i]<<endl;
    }
}